Library/OcDebugLogLib/OcLog.c

```diff
@@ -140,18 +146,129 @@ GetLogPath (
   return LogPath;
 }
 
+STATIC
 EFI_STATUS
-EFIAPI
-OcLogAddEntry  (
-  IN OC_LOG_PROTOCOL    *OcLog,
-  IN UINTN              ErrorLevel,
-  IN CONST CHAR8        *FormatString,
-  IN VA_LIST            Marker
+GetLogPrefix (
+  IN   CONST CHAR8  *FormatString,
+  OUT  CHAR8        *Prefix
   )
 {
-  EFI_STATUS                  Status;
+  UINTN    MaxLength;
+  UINTN    Index;
+  CHAR8    Curr;
 
-  OC_LOG_PRIVATE_DATA         *Private;
+  ASSERT (FormatString != NULL);
+  ASSERT (Prefix != NULL);
+
+  //
+  // If FormatString just starts with colon, it must be illegal.
+  //
+  if (*FormatString == ':') {
+    return EFI_NOT_FOUND;
+  }
+
+  MaxLength = MIN (AsciiStrLen (FormatString), OC_LOG_PREFIX_CHAR_MAX);
+  for (Index = 1; Index < MaxLength; ++Index) {
+    Curr = FormatString[Index];
+
+    //
+    // Match the first occurrence of colon.
+    //
+    if (Curr == ':') {
+      break;
+    }
+
+    //
+    // Except for colon, a valid prefix must be either 0-9, or uppercase letter.
+    //
+    if (!(IsAsciiNumber (Curr) || (Curr >= 'A' && Curr <= 'Z'))) {
+      return EFI_NOT_FOUND;
+    }
+  }
+
+  //
+  // If Index went through the end, then ':' was not found.
+  //
+  if (Index == MaxLength) {
+    return EFI_NOT_FOUND;
+  }
+
+  CopyMem (Prefix, FormatString, Index);
+  Prefix[Index] = '\0';
+  return EFI_SUCCESS;
+}
+
+STATIC
+BOOLEAN
+IsPrefixFiltered (
+  IN   CONST CHAR8          *FormatString,
+  IN   CONST OC_FLEX_ARRAY  *FlexFilters,
+  OUT  OC_LOG_FILTER_MODE   *FilterMode
+  )
+{
+  UINTN       Index;
+  CHAR8       Prefix[OC_LOG_PREFIX_CHAR_MAX + 1];
+  EFI_STATUS  Status;
+  CHAR8       **Value;
+
+  ASSERT (FormatString != NULL);
+  ASSERT (FilterMode != NULL);
+
+  if (FlexFilters == NULL) {
+    *FilterMode = OcLogFilterModeNoFilter;
+    return FALSE;
+  }
+
+  Status = GetLogPrefix (FormatString, Prefix);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCL: Failed to get log prefix - %r\n", Status));
+    return FALSE;
+  }
+
+  //
+  // Determine filter mode and process Index 0.
+  //
+  Value = (CHAR8 **) OcFlexArrayItemAt (FlexFilters, 0);
+  ASSERT (Value != NULL);
+  if ((*Value)[0] == '+') {
+    *FilterMode = OcLogFilterModePositive;
+  } else if ((*Value)[0] == '-') {
+    *FilterMode = OcLogFilterModeNegative;
+  } else {
+    *FilterMode = OcLogFilterModeNoFilter;
+  }
+
+  if (*FilterMode == OcLogFilterModeNoFilter) {
+    return FALSE;
+  }
+
+  if (AsciiStrCmp (Prefix, &((*Value)[1])) == 0) {
```

Comment:
Shall we actually check the length of `*Value` just to prevent index-out-of-bound access?

Reply:
https://github.com/acidanthera/OpenCorePkg/commit/b6cd9bd435aa17cacb23a94d89a6d9c412ca0b63